architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,652 @@
+        }
+      }
+    }
+  parameters: |
+    storageAccountName=myaccount
+```
+
+## Design
+
+### High Level Design
+<!--
+High level overview of the data flow and key components.
+
+Provide a high-level description, using diagrams as appropriate, and top-level
+explanations to convey the architectural/design overview. Don’t go into a lot
+of details yet but provide enough information about the relationship between
+these components and other components. Call out or highlight new components
+that are not part of this feature (dependencies). This diagram generally
+treats the components as black boxes. Provide a pointer to a more detailed
+design document, if one exists. 
+-->
+
+This design proposes the creation of a new Kubernetes controller (BicepDeployment controller) packaged within the `radius-controller` service.
+
+The BicepDeployment controller will be implemented as a Kubernetes controller that watches and reconciles BicepDeployment resources on the cluster. It will be responsible for compiling Bicep templates and parameters into ARM JSON and deploying the resources defined in the Bicep template to Radius. It will also be responsible for updating and deleting resources defined in the Bicep template.
+
+### Architecture Diagram
+<!--
+Provide a diagram of the system architecture, illustrating how different
+components interact with each other in the context of this proposal.
+
+Include separate high level architecture diagram and component specific diagrams, wherever appropriate.
+-->
+
+![Architecture Diagram](2024-08-bicepdeployment-controller/architecture.png)
+
+### Detailed Design
+
+<!--
+This section should be detailed and thorough enough that another developer
+could implement your design and provide enough detail to get a high confidence
+estimate of the cost to implement the feature but isn’t as detailed as the 
+code. Be sure to also consider testability in your design.
+
+For each change, give each "change" in the proposal its own section and
+describe it in enough detail that someone else could implement it. Cover
+ALL of the important decisions like names. Your goal is to get an agreement
+to proceed with coding and PRs.
+
+If there are alternatives you are considering please include that in the open
+questions section. If the product has a layered architecture, it's good to
+align these sections with the product's layers. This will help readers use
+their current understanding to understand your ideas.
+
+Discuss the rationale behind architectural choices and alternative options 
+considered during the design process.
+-->
+
+#### BicepDeployment Custom Resource Definition
+
+The `BicepDeployment` resource will be a new Kubernetes CRD that will be used to contain the Bicep template and parameters for deploying Radius resources. The CRD will have the following fields:
+
+```go
+// BicepDeploymentSpec defines the desired state of a BicepDeployment
+type BicepDeploymentSpec struct {
```

Comment:
You may also need to include the user's `bicepconfig.json` files if they have any that are picked up in the compilation. This is something we don't currently represent as "part of" the compilation, but probably should be...

Reply:
Is this relevant if the `BicepDeploymentSpec` contains ARM-JSON instead of Bicep? I think we changed the design since you left this comment, but I wanted to be sure.